File: src/main/kotlin/com/ribafish/slacksummarizer/services/GeminiService.kt
package com.ribafish.slacksummarizer.services

import com.google.ai.client.generativeai.GenerativeModel
import com.ribafish.slacksummarizer.config.GeminiConfig
import com.ribafish.slacksummarizer.models.SlackThread
import mu.KotlinLogging

private val logger = KotlinLogging.logger {}

class GeminiService(private val config: GeminiConfig) {
    private val model = GenerativeModel(
        modelName = config.model,
        apiKey = config.apiKey
    )

    suspend fun summarize(thread: SlackThread): String {
        logger.debug { "Generating summary for thread with ${thread.messages.size} messages" }

        val threadContent = buildThreadContent(thread)
        val prompt = buildPrompt(threadContent)

        logger.debug { "Prompt: $prompt" }

        val response = model.generateContent(prompt)
        val summary = response.text ?: "Failed to generate summary"

        logger.debug { "Generated summary: $summary" }

        return summary
    }

    private fun buildThreadContent(thread: SlackThread): String {
        return thread.messages.joinToString("\n\n") { message ->
            "**${message.username}** (${message.timestamp}):\n${message.text}"
        }
    }

    private fun buildPrompt(threadContent: String): String {
        return """
            You are a technical documentation assistant. Your task is to summarize a Slack conversation thread
            into a clear, well-structured markdown document suitable for a knowledge base.

            Requirements:
            1. Create a clear, descriptive title based on the main topic
            2. Write a concise summary paragraph
            3. Organize key points into logical sections
            4. Include important decisions, action items, and conclusions
            5. Preserve technical details and code snippets if present
            6. Use proper markdown formatting (headers, lists, code blocks, etc.)
            7. Keep the tone professional and clear

            Format the output as a complete markdown document with:
            - A title (# heading)
            - A summary section
            - Relevant subsections organizing the content
            - Any important links or references mentioned

            Here is the Slack thread to summarize:

            $threadContent

            Generate the markdown summary:
        """.trimIndent()
    }
}

File: summarizer-python/services/gemini_service.py
"""Google Gemini API integration service."""

import logging

import google.generativeai as genai

from ..config import GeminiConfig
from ..models import SlackThread

logger = logging.getLogger(__name__)


class GeminiService:
    """Service for generating summaries using Google Gemini."""

    def __init__(self, config: GeminiConfig):
        """Initialize Gemini service with configuration."""
        self.config = config
        genai.configure(api_key=config.api_key)
        self.model = genai.GenerativeModel(config.model)

    def summarize(self, thread: SlackThread) -> str:
        """Generate a summary of the thread using Gemini.

        Args:
            thread: The Slack thread to summarize

        Returns:
            Markdown-formatted summary
        """
        logger.debug(f"Generating summary for thread with {len(thread.messages)} messages")

        thread_content = self._build_thread_content(thread)
        prompt = self._build_prompt(thread_content)

        logger.debug("Sending request to Gemini API")

        try:
            response = self.model.generate_content(prompt)
            summary = response.text

            logger.debug(f"Generated summary: {len(summary)} characters")
            return summary

        except Exception as e:
            logger.error(f"Error calling Gemini API: {e}")
            raise

    def _build_thread_content(self, thread: SlackThread) -> str:
        """Build thread content string from messages."""
        return "\n\n".join(message.text for message in thread.messages)

    def _build_prompt(self, thread_content: str) -> str:
        """Build the prompt for Gemini."""
        return f"""You are a technical documentation assistant. Your task is to transform a Slack conversation
into a clear, well-structured knowledge base article in markdown format.

Requirements:
1. Create a clear, descriptive title based on the main topic discussed
2. Identify 3-7 keywords/tags that represent the main topics, technologies, or concepts discussed
3. Focus entirely on the technical content, concepts, and information shared
4. Organize information into logical sections with proper headers
5. Extract and preserve:
   - Technical explanations and concepts
   - Code snippets and examples
   - Solutions to problems
   - Best practices and recommendations
   - Important links and references
6. Remove conversational elements (greetings, acknowledgments, "thanks", etc.)
7. Synthesize multiple related points into cohesive explanations
8. Use proper markdown formatting (headers, lists, code blocks, etc.)
9. Keep the tone professional and encyclopedic

Do NOT include:
- Who said what or when
- Conversational back-and-forth
- Off-topic discussion
- Personal opinions unless they represent technical best practices

Format the output as a complete markdown document with:
- A descriptive title (# heading)
- A keywords line IMMEDIATELY after the title in the format: **Keywords:** keyword1, keyword2, keyword3
- An overview section explaining what this article covers
- Logical subsections organizing the technical content
- Code blocks for any code examples
- Links to external resources if mentioned

Example format:
# How to Configure Redis for High Availability

**Keywords:** redis, high-availability, clustering, replication, failover

## Overview
[content here]

Here is the conversation to transform into a knowledge base article:

{thread_content}

Generate the knowledge base article:"""
